### PR TITLE
Holdout Dataset Spike

### PR DESCRIPTION
**Pull request by christopherbunn**
Summary of changes:
- Holdout set is taken by default when dataset exceeds row minimum (currently set at 500 rows)
   - If it does not meet row minimum, then no holdout set is taken 
   - Score of holdout set is represented as `validation_score` column. Previously this row repeated the results of the `mean_cv_score` column unless only one CV fold is used (then `mean_cv_score` shows as `NaN`)
- Holdout set can also be specified and passed through `AutoMLSearch`
- `validation_score` column is moved before `mean_cv_score` in results table 
- Additional tests + test updates (typically to specify `problem_type` parameter)

TODO:
- Update docs:
   - Update start page to indicate holdout set is used

### CODE REVIEW

File: evalml/automl/pipeline_search_plots.py [hunk 1]
```diff
@@ -47,7 +47,7 @@ def update(self, results, objective):
         if len(results["search_order"]) > 0 and len(results["pipeline_results"]) > 0:
             iter_idx = results["search_order"]
             pipeline_res = results["pipeline_results"]
-            iter_scores = [pipeline_res[i]["mean_cv_score"] for i in iter_idx]
+            iter_scores = [pipeline_res[i]["validation_score"] for i in iter_idx]
```
Comment (christopherbunn) [inline evalml/automl/pipeline_search_plots.py:50]: Moving towards plotting the validation score rather than the CV score for the search iteration plot.

File: evalml/tests/automl_tests/test_automl_search_classification.py [hunk 2]
```diff
@@ -948,10 +948,10 @@ def test_automl_search_ratio_overrides_sampler_ratio(
...
     "problem_type,sampling_ratio_dict,length",
     [
-        ("binary", {0: 0.5, 1: 1}, 600),
-        ("binary", {0: 0.2, 1: 1}, 800),
-        ("multiclass", {0: 0.5, 1: 1, 2: 1}, 400),
-        ("multiclass", {0: 0.75, 1: 1, 2: 1}, 333),
+        ("binary", {0: 0.5, 1: 1}, 810),
+        ("binary", {0: 0.2, 1: 1}, 1080),
+        ("multiclass", {0: 0.5, 1: 1, 2: 1}, 540),
+        ("multiclass", {0: 0.75, 1: 1, 2: 1}, 450),
```
Comment (christopherbunn) [inline evalml/tests/automl_tests/test_automl_search_classification.py:954]: Updated values to account for the holdout set.

File: evalml/tests/automl_tests/test_automl.py [hunk 3]
```diff
@@ -4406,11 +4435,12 @@ def test_automl_passes_known_in_advance_pipeline_parameters_to_all_pipelines(
...
     data_splitter,
     mean_cv_is_none,
-    X_y_binary,
     dummy_classifier_estimator_class,
     AutoMLTestEnv,
 ):
-    X, y = X_y_binary
+    X, y = datasets.make_classification(
+        n_samples=500, n_features=20, n_informative=2, n_redundant=2, random_state=0
+    )
```
Comment (christopherbunn) [inline evalml/tests/automl_tests/test_automl.py:4443]: Bumped up size so that holdout set is taken.

File: evalml/automl/engine/engine_base.py [hunk 4]
```diff
@@ -304,26 +328,105 @@ def train_and_score_pipeline(
...
+
+    training_time = time.time() - start
     return {
         "scores": {
             "cv_data": cv_data,
             "training_time": training_time,
             "cv_scores": cv_scores,
             "cv_score_mean": cv_score_mean,
+            "holdout_score": None
+            if X_holdout is None and y_holdout is None
```
Comment (eccabay) [inline evalml/automl/engine/engine_base.py:415]: We end up checking `if X_holdout is (not) None and y_holdout is (not) None` a lot in this function, and this part is a little hard to read - could you clean it up a little by saving a local variable, `use_holdout` or something along those lines?

File: evalml/tests/automl_tests/test_automl.py [hunk 5]
```diff
@@ -702,8 +709,10 @@ def test_invalid_data_splitter(X_y_binary):
 
 
 def test_large_dataset_binary(AutoMLTestEnv):
-    X = pd.DataFrame({"col_0": [i for i in range(101000)]})
-    y = pd.Series([i % 2 for i in range(101000)])
+    X = pd.DataFrame(
+        {"col_0": [i for i in range(111113)]}
```
Comment (eccabay) [inline evalml/tests/automl_tests/test_automl.py:713]: Nitpick: 100% not a you thing (pretty sure I do the same thing), but we can simplify `[i for i in range(x)]` to just be `range(x)`

File: evalml/tests/automl_tests/test_automl.py [hunk 6]
```diff
@@ -4819,3 +4852,88 @@ def test_default_algorithm_uses_n_jobs(X_y_binary, AutoMLTestEnv):
...
+    X, y = datasets.make_classification(
+        n_samples=AutoMLSearch._HOLDOUT_SET_MIN_ROWS, random_state=0
+    )
+    automl = AutoMLSearch(
+        X_train=X, y_train=y, problem_type="binary", verbose=True, _holdout_set_size=0
+    )
+    out = caplog.text
+
+    expected_holdout_size = int(automl._holdout_set_size * len(X))
+    expected_train_size = int((1 - automl._holdout_set_size) * len(X))
```
Comment (eccabay) [inline evalml/tests/automl_tests/test_automl.py:4932]: We don't need these here

**Review by eccabay — COMMENTED** [review]
This is awesome, I'm excited for this change! I'm really curious to see how this changes our looking glass results.

File: evalml/automl/automl_search.py [hunk 7]
```diff
@@ -610,9 +625,43 @@ def __init__(
...
+                )
 
+        else:
+            self.X_train = X_train
+            self.y_train = y_train
+            self.X_holdout = X_holdout if X_holdout is not None else None
+            self.y_holdout = y_holdout if y_holdout is not None else None
+            self.logger.info(
+                f"Holdout set evaluation is disabled. AutoMLSearch will use mean CV score to rank pipelines."
+            )
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/automl/automl_search.py:673]: Updated so that message is only raised if `self.passed_holdout_set is False`

File: evalml/automl/automl_search.py [hunk 8]
```diff
@@ -610,9 +625,43 @@ def __init__(
...
+        self.X_train = infer_feature_types(self.X_train)
+        self.y_train = infer_feature_types(self.y_train)
+        if self.X_holdout is not None and self.y_holdout is not None:
+            self.X_holdout = infer_feature_types(self.X_holdout)
+            self.y_holdout = infer_feature_types(self.y_holdout)
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/automl/automl_search.py:678]: Updated else branch to use suggestion.

File: evalml/tests/automl_tests/test_automl.py (hunk 5 above)
Comment (christopherbunn) [reply evalml/tests/automl_tests/test_automl.py:751]: Moved check to `use_holdout`

File: evalml/automl/engine/engine_base.py (hunk 4 above)
Comment (christopherbunn) [reply evalml/automl/engine/engine_base.py:415]: Moved check to `use_holdout` var

File: evalml/automl/engine/engine_base.py [hunk 9]
```diff
@@ -304,26 +328,105 @@ def train_and_score_pipeline(
...
+                nan_scores = {objective: np.nan for objective in e.exceptions}
+                holdout_scores = {**nan_scores, **e.scored_successfully}
+                holdout_scores = OrderedDict(
+                    {
+                        o.name: holdout_scores[o.name]
+                        for o in [automl_config.objective]
+                        + automl_config.additional_objectives
+                    }
+                )
+                holdout_score = holdout_scores[automl_config.objective.name]
```
[the remark this answers is not in the document]
Comment (jeremyliweishih) [reply evalml/automl/engine/engine_base.py:395]: @christopherbunn what would happen if you mock both the holdout set evaluation and the CV evaluation with the same scores? Still working through in my head how that would prevent coverage here. It seems to me that it should work but may be a little confusing given that scores are the same across the board.

**Review by jeremyliweishih — CHANGES_REQUESTED** [review]
@christopherbunn great work on this and love the direction where we're heading with this. Overall things look great but I left some comments about input validation and logging. Another concern I had is that there seems to be a mix of usage between `validation data/score` and `holdout data/score`. Let's try to choose one of them for consistency and clarity. I'll give the final approval once these questions/changes are addressed!

File: evalml/automl/engine/engine_base.py [hunk 10]
```diff
@@ -304,26 +328,105 @@ def train_and_score_pipeline(
...
     cv_scores = pd.Series([fold["mean_cv_score"] for fold in cv_data])
     cv_score_mean = cv_scores.mean()
     logger.info(
         f"\tFinished cross validation - mean {automl_config.objective.name}: {cv_score_mean:.3f}"
     )
+
+    holdout_score = np.NaN
+    holdout_scores = np.NaN
+    if X_holdout is not None and y_holdout is not None:
+        logger.info("\tStarting holdout set scoring")
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/automl/engine/engine_base.py:341]: Yep, it prints for every single pipeline. I figured that since we're printing out the CV scores, it would be useful to print out the holdout scores as well. This is especially true since we're using this score as part of the search iteration plot. 

image

[the remark this answers is not in the document]
Comment (jeremyliweishih) [reply evalml/automl/engine/engine_base.py:341]: gotcha - looks way better than I expected. Keep it in!

File: evalml/automl/automl_search.py [hunk 11]
```diff
@@ -610,9 +625,43 @@ def __init__(
         self._best_pipeline = None
         self._searched = False
 
-        self.X_train = infer_feature_types(X_train)
-        self.y_train = infer_feature_types(y_train)
+        if self._holdout_set_size > 0 and self.passed_holdout_set is False:
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/automl/automl_search.py:645]: Added check on line 668 to check if `self._holdout_set_size < 0`.

File: evalml/automl/engine/engine_base.py [hunk 12]
```diff
@@ -187,18 +199,30 @@ def train_and_score_pipeline(
...
             original_target: encoded_target
             for (encoded_target, original_target) in enumerate(
                 full_y_train.value_counts().index
             )
         }
-        full_y_train = ww.init_series(full_y_train.map(y_mapping))
-    cv_pipeline = pipeline
+        full_y_train = ww.init_series(full_y_train.map(train_y_mapping))
+
+        if y_holdout is not None:
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/automl/engine/engine_base.py]: Changed to `if use_holdout`

File: evalml/automl/engine/engine_base.py [hunk 13]
```diff
@@ -304,26 +329,101 @@ def train_and_score_pipeline(
...
+                holdout_score = holdout_scores[automl_config.objective.name]
+            else:
+                holdout_score = np.nan
+                holdout_scores = OrderedDict(
+                    zip(
+                        [n.name for n in automl_config.additional_objectives],
+                        [np.nan] * len(automl_config.additional_objectives),
+                    )
+                )
+        logger.info(
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/automl/engine/engine_base.py:404]: See other comment above, but I think there's a value add in having this information printed out in verbose

File: evalml/automl/engine/engine_base.py [hunk 14]
```diff
@@ -315,26 +348,107 @@ def train_and_score_pipeline(
...
+            if (
+                automl_config.optimize_thresholds
+                and is_binary(automl_config.problem_type)
+                and full_pipeline.threshold is not None
+            ):
+                logger.debug(
+                    f"\t\t\tFull data pipeline: Optimal threshold found ({full_pipeline.threshold:.3f})",
+                )
+            logger.debug(f"\t\t\tScoring trained full training data pipeline")
+            holdout_scores = full_pipeline.score(
```
Comment (fjlanasa) [inline evalml/automl/engine/engine_base.py:383]: Curious what is `objectives_to_score` here

File: evalml/automl/automl_search.py [hunk 15]
```diff
@@ -624,9 +644,48 @@ def __init__(
...
+                )
+            else:
+                self.X_train = X_train
+                self.y_train = y_train
+                self.X_holdout = None
+                self.y_holdout = None
+                self.logger.info(
+                    f"Dataset size is too small to create holdout set. Mininum dataset size is {self._HOLDOUT_SET_MIN_ROWS} rows, X_train has {len(self.X_train)} rows. AutoMLSearch will use mean CV score to rank pipelines.",
+                )
+        elif self._holdout_set_size < 0:
```
Comment (fjlanasa) [inline evalml/automl/automl_search.py:670]: Is it possible to set holdout size to 0?

Comment (fjlanasa) [reply evalml/automl/automl_search.py:670]: Ah nvm, I see it below.

File: evalml/automl/automl_search.py [hunk 16]
```diff
@@ -454,27 +459,41 @@ def __init__(
         sampler_method="auto",
         sampler_balanced_ratio=0.25,
         allow_long_running_models=False,
-        _ensembling_split_size=0.2,
         _pipelines_per_batch=5,
         automl_algorithm="default",
         engine="sequential",
         verbose=False,
         timing=False,
+        _holdout_set_size=0.1,
```
Comment (fjlanasa) [inline evalml/automl/automl_search.py:467]: does it make sense to make the default behavior the current behavior? (setting to 0)

File: evalml/automl/engine/engine_base.py [hunk 17]
```diff
@@ -315,26 +348,107 @@ def train_and_score_pipeline(
...
+                )
+                holdout_score = holdout_scores[automl_config.objective.name]
+            else:
+                holdout_score = np.nan
+                holdout_scores = OrderedDict(
+                    zip(
+                        [n.name for n in automl_config.additional_objectives],
+                        [np.nan] * len(automl_config.additional_objectives),
+                    ),
+                )
```
[the remark this answers is not in the document]
Comment (jeremyliweishih) [reply evalml/automl/engine/engine_base.py:422]: @christopherbunn I agree with this! do you think we can refactor and use it for both CV and holdout?

File: evalml/automl/automl_search.py (hunk 16 above)
Comment (chukarsten) [reply evalml/automl/automl_search.py:467]: Yea, I was kind of getting at this to some degree in my comment about refactoring some of the nested conditional logic.  I think doing this will probably necessitate looking into how `split_data()` handles accepting a 0.  I think Chris wants it to return a None in the case that the split is empty.

File: evalml/automl/engine/engine_base.py (hunk 14 above)
Comment (christopherbunn) [reply evalml/automl/engine/engine_base.py]: Might be more clear with the new private function reorg, but it should be the main objective + any other additional objectives specified by `automl.additional_objectives`.

File: evalml/automl/engine/engine_base.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/automl/engine/engine_base.py]: Yep, I agree it's getting a bit wordy. Tried to separate it out into the `_train_and_score()` private function, lmk what you think.

File: evalml/automl/engine/engine_base.py [hunk 18]
```diff
@@ -198,18 +218,30 @@ def train_and_score_pipeline(
...
+                original_target: encoded_target
+                for (encoded_target, original_target) in enumerate(
+                    y_holdout.value_counts().index,
+                )
+            }
+            y_holdout = ww.init_series(y_holdout.map(holdout_y_mapping))
+
     pipeline_cache = {}
+    cv_pipeline = pipeline
+    stored_pipeline = cv_pipeline
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/automl/engine/engine_base.py]: Ended up removing `cv_pipeline` when separating out into `_train_and_score()` private function.

File: evalml/automl/engine/engine_base.py [hunk 19]
```diff
@@ -198,18 +218,30 @@ def train_and_score_pipeline(
...
-    cv_pipeline = pipeline
+        full_y_train = ww.init_series(full_y_train.map(train_y_mapping))
+
+        if use_holdout:
+            holdout_y_mapping = {
+                original_target: encoded_target
+                for (encoded_target, original_target) in enumerate(
+                    y_holdout.value_counts().index,
+                )
+            }
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/automl/engine/engine_base.py]: Separated out into `_encode_classification_target()`

File: evalml/automl/automl_search.py [hunk 20]
```diff
@@ -622,9 +642,48 @@ def __init__(
...
+            )
+            self.y_holdout = (
+                infer_feature_types(y_holdout) if y_holdout is not None else None
+            )
+            if self.passed_holdout_set is False:
+                self.logger.info(
+                    f"Holdout set evaluation is disabled. AutoMLSearch will use mean CV score to rank pipelines."
+                )
+        self.X_train = infer_feature_types(self.X_train)
+        self.y_train = infer_feature_types(self.y_train)
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/automl/automl_search.py]: I streamlined this a bit, I think it should be a bit more readable. I grouped together the logger messages as well as minimized the amount of feature inference. 

I also made it so that if `split_data` has a test size of 0, it returns the data as-is.

File: evalml/automl/engine/engine_base.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (christopherbunn) [reply evalml/automl/engine/engine_base.py]: With the new streamlined evaluation method, solved getting coverage for this

File: evalml/automl/automl_search.py [hunk 21]
```diff
@@ -624,8 +648,48 @@ def __init__(
         self._best_pipeline = None
         self._searched = False
 
-        self.X_train = infer_feature_types(X_train)
-        self.y_train = infer_feature_types(y_train)
+        if self.holdout_set_size < 0:
```
Comment (eccabay) [inline evalml/automl/automl_search.py:651]: We should probably also check `holdout_set_size !> 1` as well

File: evalml/automl/automl_search.py [hunk 22]
```diff
@@ -624,8 +648,48 @@ def __init__(
...
+            raise ValueError(
+                "Holdout set size must be greater than 0. Set holdout set size to 0 to disable holdout set evaluation.",
+            )
+        if (
```
Comment (eccabay) [inline evalml/automl/automl_search.py:655]: I had a bit of trouble understanding what each case here does, could you add some comments to make it easier to read?

File: evalml/automl/automl_search.py (hunk 16 above)
Comment (christopherbunn) [reply evalml/automl/automl_search.py]: Ended up bumping `holdout_set_size` to public and will keep as default/

File: evalml/automl/automl_search.py [hunk 23]
```diff
@@ -624,8 +652,50 @@ def __init__(
         self._best_pipeline = None
         self._searched = False
 
-        self.X_train = infer_feature_types(X_train)
-        self.y_train = infer_feature_types(y_train)
+        if self.holdout_set_size < 0 or holdout_set_size >= 1:
```
Comment (jeremyliweishih) [inline evalml/automl/automl_search.py:655]: I'm finding the flow of logic in these if statements to be confusing, would this following refactor cover all the bases?  Happy to run through this together to clarify these if blocks.
```
if self.holdout_set_size < 0 or holdout_set_size >= 1: # first validate if we're doing holdout

elif self.passed_holdout_set is False: # then check if holdout is not passed
     if len(X_train) >= self._HOLDOUT_SET_MIN_ROWS): # if > 500 create and log we're doing holdout
               ...
     else: # if 500 log and do nothing
             ....
else: # we're using passed in holdout set - log we're doing holdout.
```

File: evalml/automl/engine/engine_base.py [hunk 24]
```diff
@@ -196,78 +216,54 @@ def train_and_score_pipeline(
...
-            diff_string += (
-                f"Missing target values in the validation set after data split: {diff_valid}."
-                if diff_valid
-                else ""
-            )
-            if diff_string:
-                raise Exception(diff_string)
+        return ww.init_series(y.map(y_mapping))
+
+    def _train_and_score(X_train, X_score, y_train, y_score, fold_num=None):
```
Comment (jeremyliweishih) [inline evalml/automl/engine/engine_base.py:229]: this looks so clean 🤩

**Review by jeremyliweishih — CHANGES_REQUESTED** [review]
Looks great! Just have some final doc thoughts and a potential refactor on the if logic for creating the holdout set.

**Review by chukarsten — APPROVED** [review]
Love it!  All good changes.  Thanks for taking the time, Chris.

File: docs/source/user_guide/automl.ipynb [hunk 25]
```diff
@@ -137,6 +137,24 @@
...
+   "metadata": {},
+   "source": [
+    "### Holdout Set for Pipeline Ranking"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "For datasets that have more than 500 rows, AutoMLSearch will automatically create a holdout set from 10% of the training data. Alternatively, a holdout set can be manually specified by using the `X_holdout` and `y_holdout` parameters in `AutoMLSearch()`.\n",
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply docs/source/user_guide/automl.ipynb]: Let's pass it in here, updated.

File: docs/source/user_guide/automl.ipynb [hunk 26]
```diff
@@ -137,6 +137,24 @@
...
+    "\n",
+    "During the AutoML search process, the pipeline will be fitted on the entire training dataset and scored on this new holdout set. This score is represented under the \"validation_score\" column on the pipeline rankings board. \n",
+    "\n",
+    "If a dataset has less than 500 rows or `holdout_set_size=0`, the mean of the objective scores of all cross validation folds (shown the \"mean_cv_score\" column in the pipeline rankings) will be used as the validation_score instead."
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply docs/source/user_guide/automl.ipynb]: Added some more info, hopefully it's more clear?

File: docs/source/user_guide/automl.ipynb [hunk 27]
```diff
@@ -356,7 +374,9 @@
    "metadata": {},
    "source": [
     "## View Rankings\n",
-    "A summary of all the pipelines built can be returned as a pandas DataFrame which is sorted by score. The score column contains the average score across all cross-validation folds while the validation_score column is computed from the first cross-validation fold."
+    "A summary of all the pipelines built can be returned as a pandas DataFrame which is sorted by the validation score.\n",
+    "- For AutoML searches completed with a holdout set, the validation score is the holdout score of the pipeline fitted using the entire training dataset. \n",
```
[the remark this answers is not in the document]
Comment (christopherbunn) [reply docs/source/user_guide/automl.ipynb:386]: Tried a new fix for it, hopefully it formats better?

File: evalml/automl/automl_search.py (hunk 23 above)
Comment (christopherbunn) [reply evalml/automl/automl_search.py:655]: I think I adapted most of these changes. I left a separate log block below that prints out what the validation score is for clarity.

**Review by jeremyliweishih — APPROVED** [review]

**Comment by jeremyliweishih** [comment]
@christopherbunn LGTM!
